doc/fbrun.1
.\"
.\" Man page for fbrun
.\"
.\" Copyright (c) 2004 by Bastian Kleineidam <reflexionsniveau@web.de>
.\"
.TH FBRUN 1 "16 December 2003"
.SH NAME
fbrun \- display run dialog window
.SH SYNOPSIS
\fBfbrun\fP [\fIoptions\fP]...
.SH DESCRIPTION
\fBfbrun\fP is basically equivalent to the "Run..." dialog in other
desktop environments. This means that it is an easy way to start a
program that isn't contained in the menu (or needs a special set of
parameters for this particular invocation).

Another way fbrun can be useful is to be called from the menu with a
preloaded command line that you can edit and then execute. An example
might be sshing to a very long host name with lots of options of which
one changes all the time. In this case, you could add an entry for fbrun
to your menu that contains all the options and the host name. When you use
said entry, you could edit the line as necessary and execute it.
.SH OPTIONS
.TP
\fB-font\fP \fIname\fP
Text font name
.TP
\fB-title\fP \fIname\fP
Set title
.TP
\fB-text\fP \fItext\fP
Text input
.TP
\fB-w\fP \fIwidth\fP
Window width in pixels
.TP
\fB-h\fP \fIheight\fP
Window height in pixels
.TP
\fB-display\fP \fIname\fP
Display name
.TP
\fB-pos\fP \fIx\fP \fIy\fP
Window position in pixels
.TP
\fB-nearmouse\fP
Position the window under the mouse cursor
.TP
\fB-fg\fP \fIname\fP
Foreground text color
.TP
\fB-bg\fP \fIname\fP
Background color
.TP
\fB-na\fP
Disable antialias
.TP
\fB-hf\fP \fIfilename\fP
History file to load (default ~/.fluxbox/fbrun_history
.TP
\fB-help\fP
Show this help
.SH EXAMPLE
  \fBfbrun -fg black -bg white -text xterm -title "run xterm"\fP
.SH AUTHOR
This manpage was originally written by Bastian Kleineidam
<calvin@debian.org> for the Debian distribution of fluxbox
(but may be used by others).

The main author of fluxbox is Henrik Kinnunen <fluxgen@fluxbox.org>.


NEWS
News in 1.0.0:
    * New default style: bloe
    * New styles:
      ostrich zimek_darkblue, zimek_green, zimek_bisque,
      carp, arch, bora_black, bora_blue, bora_green, green_tea, bloe
    * Fixed shape handling, now shaped corners have border.
    * Allow negative numbers in command "Workspace"
    * Added utility fluxbox-remote 
      * Allows script access to most fluxbox commands; 
        Support must be turned on in the init file with
        session.screen<N>.allowRemoteActions; the syntax is, e.g.,
        fluxbox-remote "CustomMenu ~/.fluxbox/custommenu" 
    * Added keycommands :AddWorkspace and :RemoveLastWorkspace
    * Added some missing Open Office entries and window managers,
      plus some other minor changes in fluxbox-generate_menu
    * Updated languages: 
      sv_SE, zh_CN, de_*, es_ES, es_AR, en_US, en_GB, pt_*
      fi_FI, fr_FR, ru_RU

Bug fixes:
    * #1694959 Rotate toolbar background texture along with toolbar
    * #1712583 Non-latin characters displayed incorrectly in menu title
    * #1708859 Fix error passing --with-locale path in ./configure
    * #1524098 #1633718, System tray restart issue 
               ( gaim/pidgin crash on restart )
    * #1748967 Save workspace names when changing through menu
    * #1745678 Fixed resource usage due to Skype 1.4
    * #1736252 Strip off additional bits in event state set by Xorg
    * #1716899 Deiconify when automatically tabbing to a minimized 
               window with focus new windows on
    * #1479517 Fix submenu placement with xinerama after moving the 
               menu between heads
    * #1716900 Also send transient windows when sending a window to 
               another workspace
    * #1717782 Bad fullscreen in ooimpress2
    * #1718112 memory leak in FbWindow::textProperty 
    * #1545066 Only replace numbers with zeros when calculating clock 
               width
    * #1652185 Fix window menu context when opened by clicking on tab
    * Workspace Menu wasn't updating when windows changed titles 
    * HideMenus didn't work on CustomMenus, and also fixed a memleak with
      CustomMenu
    * Fix to avoid rogue instances of /bin/sh after forking away programs
    * Fixed some rendering problems when toolbar item borderwidth is
      too large and toolbar bevelwidth is set
    * Right clicking on a submenu didn't revert focus to its parent
    * Pressing escape in a torn menu didn't work
    * Updates for compiling with gcc 4.3 
    * Fixed compile error with --disable-remember
    * Fix transient windows getting tabbed from apps file and not 
      changing layers when main window is in a lowered tab
    * Fixed some window placement issues
    * Fixed bug with fluxbox-generate_menu -is
    * Fixed infinite loop caused by deiconify command 
    * Fixed a bug in parsing cli
    * Fix slit.onhead with BOTTOMLEFT placement
    * Fixed crash when CustomMenu command doesn't point to a file

See ChangeLog for more details.




util/fbrun/FbRun.cc
// FbRun.cc
// Copyright (c) 2002 - 2006 Henrik Kinnunen (fluxgen at fluxbox dot org)
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#include "FbRun.hh"

#include "App.hh"
#include "EventManager.hh"
#include "Color.hh"
#include "KeyUtil.hh"
#include "FileUtil.hh"

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif // HAVE_CONFIG_H

#ifdef HAVE_XPM
#include <X11/xpm.h>
#include "fbrun.xpm"
#endif // HAVE_XPM

#include <X11/Xlib.h>
#include <X11/keysym.h>
#include <X11/Xutil.h>
#include <X11/cursorfont.h>
#include <unistd.h>

#include <iostream>
#include <iterator>
#include <fstream>
#include <algorithm>

using std::cerr;
using std::endl;
using std::string;
using std::fstream;
using std::ifstream;
using std::ofstream;
using std::ios;

FbRun::FbRun(int x, int y, size_t width):
    FbTk::TextBox(DefaultScreen(FbTk::App::instance()->display()),
                  m_font, ""),
    m_font("fixed"),
    m_display(FbTk::App::instance()->display()),
    m_bevel(4),
    m_gc(*this),
    m_end(false),
    m_current_history_item(0),
    m_last_completion_prefix(""),
    m_current_apps_item(0),
    m_cursor(XCreateFontCursor(FbTk::App::instance()->display(), XC_xterm)) {

    setGC(m_gc.gc());
    setCursor(m_cursor);
    // setting nomaximize in local resize
    resize(width, font().height() + m_bevel);

    // setup class name
    XClassHint *class_hint = XAllocClassHint();
    if (class_hint == 0)
        throw string("Out of memory");
    class_hint->res_name = const_cast<char *>("fbrun");
    class_hint->res_class = const_cast<char *>("FbRun");
    XSetClassHint(m_display, window(), class_hint);

    XFree(class_hint);
#ifdef HAVE_XPM
    Pixmap mask = 0;
    Pixmap pm;
    XpmCreatePixmapFromData(m_display,
                            window(),
                            const_cast<char **>(fbrun_xpm),
                            &pm,
                            &mask,
                            0); // attribs
    if (mask != 0)
        XFreePixmap(m_display, mask);

    m_pixmap = pm;
#endif // HAVE_XPM

    if (m_pixmap.drawable()) {
        XWMHints wmhints;
        wmhints.flags = IconPixmapHint;
        wmhints.icon_pixmap = m_pixmap.drawable();
        XSetWMHints(m_display, window(), &wmhints);
    }

}


FbRun::~FbRun() {
    hide();
}

void FbRun::run(const std::string &command) {
    FbTk::App::instance()->end(); // end application
    m_end = true; // mark end of processing

    // fork and execute program
    if (!fork()) {

        const char *shell = getenv("SHELL");
        if (!shell)
            shell = "/bin/sh";

        setsid();
        execl(shell, shell, "-c", command.c_str(), static_cast<void*>(NULL));
        exit(0); //exit child
    }

    hide(); // hide gui

    // save command history to file
    if (text().size() != 0) { // no need to save empty command

        // don't allow duplicates into the history file, first
        // look for a duplicate
        if (m_current_history_item < m_history.size()
            && text() == m_history[m_current_history_item]) {
            // m_current_history_item is the duplicate
        } else {
            m_current_history_item = 0;
            for (; m_current_history_item < m_history.size();
                 ++m_current_history_item) {
                if (m_history[m_current_history_item] == text())
                    break;
            }
        }

        fstream inoutfile(m_history_file.c_str(), ios::in|ios::out);
        if (inoutfile) {
            // now m_current_history_item points at the duplicate, or
            // at m_history.size() if no duplicate
            if (m_current_history_item != m_history.size()) {
                unsigned int i = 0;
                // read past history items before current
                for (; inoutfile.good() && i < m_current_history_item; i++)
                    inoutfile.ignore(1, '\n');

                // write the history items that come after current
                for (i++; i < m_history.size(); i++)
                    inoutfile<<m_history[i]<<endl;

            } else {
                // set put-pointer at end of file
                inoutfile.seekp(0, ios::end);
            }
            // append current command to the file
            inoutfile<<command<<endl;

        } else
            cerr<<"FbRun Warning: Can't write command history to file: "<<m_history_file<<endl;
    }

}

bool FbRun::loadHistory(const char *filename) {
    if (filename == 0)
        return false;
    ifstream infile(filename);
    if (!infile) {
        //even though we fail to load file, we should try save to it
        ofstream outfile(filename);
        if (outfile) {
            m_history_file = filename;
            return true;
        }
        return false;
    }
    // clear old history and load new one from file
    m_history.clear();
    // each line is a command
    string line;
    while (!infile.eof()) {
        getline(infile, line);
        if (line.size()) // don't add empty lines
            m_history.push_back(line);
    }
    // set no current histor to display
    m_current_history_item = m_history.size();
    // set history file
    m_history_file = filename;
    return true;
}

bool FbRun::loadFont(const string &fontname) {
    if (!m_font.load(fontname.c_str()))
        return false;

    // resize to fit new font height
    resize(width(), font().height() + m_bevel);
    return true;
}

void FbRun::setForegroundColor(const FbTk::Color &color) {
    m_gc.setForeground(color);
}

void FbRun::setTitle(const string &title) {
    setName(title.c_str());
}

void FbRun::resize(unsigned int width, unsigned int height) {
    FbTk::TextBox::resize(width, height);
}

void FbRun::redrawLabel() {
    clear();
}

void FbRun::keyPressEvent(XKeyEvent &ke) {
    // reset last completion prefix if we don't do a tab completion thing
    bool did_tab_complete = false;

    ke.state = FbTk::KeyUtil::instance().cleanMods(ke.state);

    FbTk::TextBox::keyPressEvent(ke);
    KeySym ks;
    char keychar[1];
    XLookupString(&ke, keychar, 1, &ks, 0);
    // a modifier key by itself doesn't do anything
    if (IsModifierKey(ks))
        return;

    if (FbTk::KeyUtil::instance().isolateModifierMask(ke.state)) { // a modifier key is down
        if ((ke.state & ControlMask) == ControlMask) {
            switch (ks) {
            case XK_p:
                did_tab_complete = true;
                prevHistoryItem();
                break;
            case XK_n:
                did_tab_complete = true;
                nextHistoryItem();
                break;
            case XK_Tab:
                did_tab_complete = true;
                tabCompleteHistory();
                break;
            }
        } else if ((ke.state & (Mod1Mask|ShiftMask)) == (Mod1Mask | ShiftMask)) {
            switch (ks) {
            case XK_less:
                did_tab_complete = true;
                firstHistoryItem();
                break;
            case XK_greater:
                did_tab_complete = true;
                lastHistoryItem();
                break;
            }
        }
    } else { // no modifier key
        switch (ks) {
        case XK_Escape:
            m_end = true;
            hide();
            FbTk::App::instance()->end(); // end program
            break;
        case XK_KP_Enter:
        case XK_Return:
            run(text());
            break;
        case XK_Up:
            prevHistoryItem();
            break;
        case XK_Down:
            nextHistoryItem();
            break;
        case XK_Tab:
            did_tab_complete = true;
            tabCompleteApps();
            break;
        }
    }
    clear();
    if (!did_tab_complete)
        m_last_completion_prefix = "";
}

void FbRun::lockPosition(bool size_too) {
    // we don't need to maximize this window
    XSizeHints sh;
    sh.flags = PMaxSize | PMinSize;
    sh.max_width = width();
    sh.max_height = height();
    sh.min_width = width();
    sh.min_height = height();
    if (size_too) {
        sh.flags |= USPosition;
        sh.x = x();
        sh.y = y();
    }
    XSetWMNormalHints(m_display, window(), &sh);
}

void FbRun::prevHistoryItem() {
    if (m_history.size() == 0 || m_current_history_item == 0) {
        XBell(m_display, 0);
    } else {
        m_current_history_item--;
        setText(m_history[m_current_history_item]);
    }
}

void FbRun::nextHistoryItem() {
    if (m_current_history_item == m_history.size()) {
        XBell(m_display, 0);
    } else {
        m_current_history_item++;
        if (m_current_history_item == m_history.size()) {
            m_current_history_item = m_history.size();
            setText("");
         } else
            setText(m_history[m_current_history_item]);
    }
}

void FbRun::firstHistoryItem() {
    if (m_history.size() == 0 || m_current_history_item == 0) {
        XBell(m_display, 0);
    } else {
        m_current_history_item = 0;
        setText(m_history[m_current_history_item]);
    }
}

void FbRun::lastHistoryItem() {
    // actually one past the end
    if (m_history.size() == 0) {
        XBell(m_display, 0);
    } else {
        m_current_history_item = m_history.size();
        setText("");
    }
}

void FbRun::tabCompleteHistory() {
    if (m_current_history_item == 0 || m_history.empty() ) {
        XBell(m_display, 0);
    } else {
        unsigned int nr= 0;
        unsigned int history_item = m_current_history_item - 1;
        if (m_last_completion_prefix.empty())
            m_last_completion_prefix = text().substr(0, textStartPos() + cursorPosition());

        while (history_item != m_current_history_item && nr++ < m_history.size()) {
            if (m_history[history_item].find(m_last_completion_prefix) == 0) {
                m_current_history_item = history_item;
                setText(m_history[m_current_history_item]);
                cursorEnd();
                break;
            }
            if (history_item == 0) // loop
                history_item = m_history.size();
            history_item--;
        }
        if (history_item == m_current_history_item) XBell(m_display, 0);
    }
}

void FbRun::tabCompleteApps() {

    static bool first_run= true;
    if (m_last_completion_prefix.empty())
        m_last_completion_prefix = text().substr(0, textStartPos() + cursorPosition());
    string prefix = m_last_completion_prefix;
    FbTk::Directory dir;

    bool add_dirs= false;
    bool changed_prefix= false;

    // (re)build m_apps-container
    if (first_run || m_last_completion_prefix != prefix) {
        first_run= false;

        string path;

        if(!prefix.empty() &&
            string("/.~").find_first_of(prefix[0]) != string::npos) {
            size_t rseparator= prefix.find_last_of("/");
            path= prefix.substr(0, rseparator + 1) +  ":";
            add_dirs= true;
        } else {
            char* tmp_path = getenv("PATH");
            if (tmp_path)
                path = tmp_path;
        }
        m_apps.clear();

        unsigned int l;
        unsigned int r;

        for(l= 0, r= 0; r < path.size(); r++) {
            if ((path[r]==':' || r == path.size() - 1) && r - l > 0) {
                string filename;
                string fncomplete;
                dir.open(path.substr(l, r - l).c_str());
                int n= dir.entries();
                if (n >= 0) {
                    while(n--) {
                        filename= dir.readFilename();
                        fncomplete= dir.name() +
                                    (*dir.name().rbegin() != '/' ? "/" : "") +
                                    filename;

                        // directories in dirmode ?
                        if (add_dirs && FbTk::FileUtil::isDirectory(fncomplete.c_str()) &&
                            filename != ".." && filename != ".") {
                            m_apps.push_back(fncomplete);
                        // executables in dirmode ?
                        } else if (add_dirs && FbTk::FileUtil::isRegularFile(fncomplete.c_str()) &&
                                   FbTk::FileUtil::isExecutable(fncomplete.c_str()) &&
                                   (prefix == "" ||
                                    fncomplete.substr(0, prefix.size()) == prefix)) {
                            m_apps.push_back(fncomplete);
                        // executables in $PATH ?
                        } else if (FbTk::FileUtil::isRegularFile(fncomplete.c_str()) &&
                                   FbTk::FileUtil::isExecutable(fncomplete.c_str()) &&
                                   (prefix == "" ||
                                    filename.substr(0, prefix.size()) == prefix)) {
                            m_apps.push_back(filename);
                        }
                    }
                }
                l= r + 1;
                dir.close();
            }
        }
        sort(m_apps.begin(), m_apps.end());
        unique(m_apps.begin(), m_apps.end());

        m_last_completion_prefix = prefix;
        changed_prefix= true;
        m_current_apps_item= 0;
    }

    if (m_apps.empty() ) {
      XBell(m_display, 0);
    } else {
        size_t apps_item = m_current_apps_item + (changed_prefix ? 0 : 1);
        bool loop= false;

        while (true) {
            if (apps_item >= m_apps.size() ) {
                loop = true;
                apps_item = 0;
            }

            if ((!changed_prefix || loop) && apps_item == m_current_apps_item) {
                break;
            }
            if (m_apps[apps_item].find(prefix) == 0) {
                m_current_apps_item = apps_item;
                if (add_dirs && FbTk::FileUtil::isDirectory(m_apps[m_current_apps_item].c_str()))
                    setText(m_apps[m_current_apps_item] +  "/");
                else
                    setText(m_apps[m_current_apps_item]);
                cursorEnd();
                break;
            }
            apps_item++;
        }
        if (!changed_prefix && apps_item == m_current_apps_item)
            XBell(m_display, 0);
    }
}

void FbRun::insertCharacter(char keychar) {
    char val[2] = {keychar, 0};
    insertText(val);
}

